src/Common/src/Interop/Unix/System.Net.Security.Native/Interop.Gss.cs

```diff
@@ -0,0 +1,336 @@
+            }
+        }
+
+        // Return a pointer to where data resides
+        public IntPtr Value
+        {
+            get
+            {
+                if (IsInvalid)
+                {
+                    return IntPtr.Zero;
+                }
+                return ((Interop.libgssapi.gss_buffer_desc)_gch.Target).value;
+            }
+        }
+
+        public SafeGssBufferHandle() : base(IntPtr.Zero, true)
+        {
+            Interop.libgssapi.gss_buffer_desc buffer = new Interop.libgssapi.gss_buffer_desc
+            {
+                length = (size_t)0,
+                value = IntPtr.Zero,
+            };
+
+            _gch = GCHandle.Alloc(buffer, GCHandleType.Pinned);
+            handle = _gch.AddrOfPinnedObject();
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        // Note that _value should never be freed directly. For input
+        // buffer, it is owned by the caller and for output buffer,
+        // it is owned by libgssapi
+        protected override bool ReleaseHandle()
+        {
+            Interop.libgssapi.gss_buffer_desc buffer = (Interop.libgssapi.gss_buffer_desc) _gch.Target;
+            if (buffer.value != IntPtr.Zero)
+            {
+                Interop.libgssapi.Status minorStatus;
+                Interop.libgssapi.Status status = Interop.libgssapi.GssReleaseBuffer(out minorStatus, ref buffer);
+                Interop.libgssapi.GssApiException.AssertOrThrowIfError("GssReleaseBuffer failed", status, minorStatus);
+            }
+
+            _gch.Free();
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a gss_name_t_desc*
+    /// </summary>
+    internal sealed class SafeGssNameHandle : SafeHandle
+    {
+        public static SafeGssNameHandle Create(string name, bool isUser)
+        {
+            Debug.Assert(!String.IsNullOrEmpty(name), "Invalid name passed to SafeGssNameHandle create");
+            SafeGssNameHandle retHandle;
+            Interop.libgssapi.Status minorStatus;
+            Interop.libgssapi.Status status = Interop.libgssapi.GssImportName(out minorStatus, name, isUser, out retHandle);
+            if (status != Interop.libgssapi.Status.GSS_S_COMPLETE)
+            {
+                throw Interop.libgssapi.GssApiException.Create(status, minorStatus);
+            }
+
+            return retHandle;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.libgssapi.Status minorStatus;
+            Interop.libgssapi.Status status = Interop.libgssapi.GssReleaseName(out minorStatus, ref handle);
+            Interop.libgssapi.GssApiException.AssertOrThrowIfError("GssReleaseName failed", status, minorStatus);
```

Comment:
Since you're explicitly setting the handle to zero, you don't want to do that in the failure case as well?